File: tools/run_albert.c
// run_albert.c — Pure C Albert-MoE-13 inference (CPU, libc only)
// Build: gcc -O3 run_albert.c -lm -o run_albert && ./run_albert "prompt"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>
#include <time.h>
#include <stdint.h>

#define W(name) ({ float *_w = get_w(m, name); if (!_w) { fprintf(stderr,"WARN: missing %s\n",name); static float _z[H]; memset(_z,0,sizeof(_z)); _w = _z; } _w; })

#define H 256
#define N_HEADS 4
#define HEAD_DIM 64
#define N_EXP 12
#define TOP_K 3
#define VOCAB 32000
#define FFN_HIDDEN 1024
#define MAX_SEQ 256

// ── Load model weights via binary index ──
typedef struct { char key[128]; float *data; int64_t numel; } Tensor;
typedef struct { Tensor *t; int count; float *raw; size_t raw_size; } Model;

static Model load_model(const char *sf_path, const char *idx_path) {
    Model m = {0};
    // Read index
    FILE *fi = fopen(idx_path, "rb"); if (!fi) { perror(idx_path); return m; }
    int count; fread(&count, 4, 1, fi);
    m.t = calloc(count, sizeof(Tensor));
    m.count = count;
    
    uint8_t *name_buf = NULL; size_t name_cap = 0;
    int64_t off, nel;
    for (int i = 0; i < count; i++) {
        int name_len; if (!fread(&name_len, 4, 1, fi)) break;
        if (name_len > (int)name_cap) { name_cap = name_len + 1; name_buf = realloc(name_buf, name_cap); }
        if (!fread(name_buf, 1, name_len, fi)) break; name_buf[name_len] = 0;
        strcpy(m.t[i].key, (char*)name_buf);
        if (!fread(&off, 8, 1, fi)) break;  // offset
        if (!fread(&nel, 8, 1, fi)) break;  // numel
        m.t[i].numel = nel;
        m.t[i].data = NULL; // will be set in second pass
    }
    free(name_buf); fclose(fi);
    printf("Read %d index entries\n", count);
    
    // Read safetensors data
    FILE *fs = fopen(sf_path, "rb"); if (!fs) { perror(sf_path); return m; }
    fseek(fs, 0, SEEK_END); m.raw_size = ftell(fs); fseek(fs, 0, SEEK_SET);
    m.raw = malloc(m.raw_size); fread(m.raw, 1, m.raw_size, fs); fclose(fs);
    
    // Map data pointers — re-read index for proper offsets
    fi = fopen(idx_path, "rb"); fread(&count, 4, 1, fi);
    for (int i = 0; i < count && i < m.count; i++) {
        int name_len; fread(&name_len, 4, 1, fi);
        fseek(fi, name_len, SEEK_CUR);
        int64_t offset, numel;
        fread(&offset, 8, 1, fi);
        fread(&numel, 8, 1, fi);
        if (i < m.count) {
            m.t[i].data = (float*)(m.raw + offset);
            m.t[i].numel = numel;
        }
    }
    fclose(fi);
    printf("Model: %d tensors, %.1f MB\n", count, m.raw_size/1e6);
    return m;
}

static float *get_w(Model *m, const char *name) {
    for (int i = 0; i < m->count; i++)
        if (!strcmp(m->t[i].key, name)) return m->t[i].data;
    fprintf(stderr, "Missing: %s\n", name); return NULL;
}

// ── Math ──
static float gelu(float x) {
    float t = tanhf(0.79788456f * (x + 0.044715f * x * x * x));
    return 0.5f * x * (1.0f + t);
}

static void layer_norm(float *x, float *w, float *b, int n) {
    float mean = 0, var = 0;
    for (int i = 0; i < n; i++) mean += x[i];
    mean /= n;
    for (int i = 0; i < n; i++) { float d = x[i] - mean; var += d * d; }
    var = sqrtf(var / n + 1e-5f);
    for (int i = 0; i < n; i++) x[i] = (x[i] - mean) / var * w[i] + b[i];
}

static void matmul_t(float *out, float *a, float *b, int m, int n, int k) {
    for (int i = 0; i < m; i++)
        for (int j = 0; j < n; j++) {
            float s = 0;
            for (int t = 0; t < k; t++) s += a[i*k+t] * b[j*k+t];
            out[i*n+j] = s;
        }
}

// ═══ Forward ───
static void forward(float *x, int seq, Model *m, int *tokens) {
    float buf[MAX_SEQ * H]; // temp buffer
    
    // Embed
    printf("DEBUG: m->count=%d raw=%p raw_size=%zu\n", m->count, (void*)m.raw, m.raw_size); fflush(stdout);
    printf("DEBUG: first key='%s'\n", m->t[0].key); fflush(stdout);
    printf("DEBUG: first data ptr=%p\n", (void*)m->t[0].data); fflush(stdout);
    float *embed_w = get_w(m, "embed.weight");
    printf("DEBUG: embed_w=%p\n", (void*)embed_w); fflush(stdout);
    if (!embed_w) { fprintf(stderr,"Missing embed.weight\n"); exit(1); }
    ptrdiff_t offset = (char*)embed_w - (char*)m.raw;
    printf("DEBUG: offset=%td numel=%ld\n", offset, (long)(m->t[0].numel)); fflush(stdout);
    if (offset < 0 || offset + H*4 > (ptrdiff_t)m.raw_size) {
        fprintf(stderr,"ERROR: embed_w out of bounds!\n"); exit(1);
    }
    printf("DEBUG: embed_w[0]=%.4f\n", embed_w[0]); fflush(stdout);
    for (int i = 0; i < seq; i++) {
        int tok = tokens[i]; if (tok < 0 || tok >= 32000) tok = 0;
        memcpy(&x[i*H], &embed_w[tok*H], H*sizeof(float));
    }
    
    // Run all blocks (both streams)
    for (int blk = 0; blk < 33; blk++) {
        char key[256];
        
        for (int stream = 0; stream < 2; stream++) {
            char s = stream ? 'b' : 'a';
            
            // ── Attention sublayer ──
            snprintf(key, sizeof(key), "blocks.%d.stream_%c.ln1.weight", blk, s);
            float *ln1_w = get_w(m, key);
            snprintf(key, sizeof(key), "blocks.%d.stream_%c.ln1.bias", blk, s);
            float *ln1_b = get_w(m, key);
            snprintf(key, sizeof(key), "blocks.%d.stream_%c.attn.q_proj.weight", blk, s);
            float *q_w = get_w(m, key);
            snprintf(key, sizeof(key), "blocks.%d.stream_%c.attn.k_proj.weight", blk, s);
            float *k_w = get_w(m, key);
            snprintf(key, sizeof(key), "blocks.%d.stream_%c.attn.v_proj.weight", blk, s);
            float *v_w = get_w(m, key);
            snprintf(key, sizeof(key), "blocks.%d.stream_%c.attn.o_proj.weight", blk, s);
            float *o_w = get_w(m, key);
            
            memcpy(buf, x, seq*H*sizeof(float));
            for (int i = 0; i < seq; i++) layer_norm(&buf[i*H], ln1_w, ln1_b, H);
            
            // Standard attention
            float q[seq*H], k[seq*H], v[seq*H], attn[seq*seq];
            matmul_t(q, buf, q_w, seq, H, H);
            matmul_t(k, buf, k_w, seq, H, H);
            matmul_t(v, buf, v_w, seq, H, H);
            
            float scale = 1.0f / sqrtf(HEAD_DIM);
            for (int h = 0; h < N_HEADS; h++) {
                for (int i = 0; i < seq; i++)
                    for (int j = 0; j < seq; j++) {
                        float s = 0;
                        for (int d = 0; d < HEAD_DIM; d++)
                            s += q[i*H+h*HEAD_DIM+d] * k[j*H+h*HEAD_DIM+d];
                        attn[i*seq+j] = s * scale;
                    }
                for (int i = 0; i < seq; i++) {
                    float mv = attn[i*seq]; for (int j = 1; j < seq; j++) if (attn[i*seq+j] > mv) mv = attn[i*seq+j];
                    float sum = 0; for (int j = 0; j < seq; j++) { attn[i*seq+j] = expf(attn[i*seq+j] - mv); sum += attn[i*seq+j]; }
                    for (int j = 0; j < seq; j++) attn[i*seq+j] /= sum;
                }
                for (int i = 0; i < seq; i++) {
                    float out_h[HEAD_DIM] = {0};
                    for (int j = 0; j < seq; j++)
                        for (int d = 0; d < HEAD_DIM; d++)
                            out_h[d] += attn[i*seq+j] * v[j*H+h*HEAD_DIM+d];
                    for (int d = 0; d < HEAD_DIM; d++) buf[i*H+h*HEAD_DIM+d] = out_h[d];
                }
            }
            // Output projection + residual
            float tmp[seq*H];
            matmul_t(tmp, buf, o_w, seq, H, H);
            for (int i = 0; i < seq*H; i++) x[i] += tmp[i];
            
            // ── MoE sublayer ──
            snprintf(key, sizeof(key), "blocks.%d.stream_%c.ln2.weight", blk, s);
            float *ln2_w = get_w(m, key);
            snprintf(key, sizeof(key), "blocks.%d.stream_%c.ln2.bias", blk, s);
            float *ln2_b = get_w(m, key);
            snprintf(key, sizeof(key), "blocks.%d.stream_%c.moe.gate.weight", blk, s);
            float *gate_w = get_w(m, key);
            
            memcpy(buf, x, seq*H*sizeof(float));
            for (int i = 0; i < seq; i++) layer_norm(&buf[i*H], ln2_w, ln2_b, H);
            
            // Router
            float gs[seq*N_EXP];
            matmul_t(gs, buf, gate_w, seq, N_EXP, H);
            
            // Expert FFN (use averaged expert weights)
            float fc_w[H*FFN_HIDDEN], fc_b[FFN_HIDDEN], proj_w[FFN_HIDDEN*H], proj_b[H];
            int exp_count = 0;
            memset(fc_w, 0, sizeof(fc_w)); memset(fc_b, 0, sizeof(fc_b));
            memset(proj_w, 0, sizeof(proj_w)); memset(proj_b, 0, sizeof(proj_b));
            
            for (int eb = 3; eb < 33; eb++) {
                for (int e = 0; e < N_EXP; e++) {
                    snprintf(key, sizeof(key), "blocks.%d.experts.%d.c_fc.weight", eb, e);
                    float *w = get_w(m, key); if (!w) continue;
                    snprintf(key, sizeof(key), "blocks.%d.experts.%d.c_fc.bias", eb, e);
                    float *bi = get_w(m, key);
                    snprintf(key, sizeof(key), "blocks.%d.experts.%d.c_proj.weight", eb, e);
                    float *pw = get_w(m, key);
                    snprintf(key, sizeof(key), "blocks.%d.experts.%d.c_proj.bias", eb, e);
                    float *pb = get_w(m, key);
                    
                    for (int j = 0; j < H*FFN_HIDDEN; j++) fc_w[j] += w[j];
                    for (int j = 0; j < FFN_HIDDEN; j++) fc_b[j] += bi[j];
                    for (int j = 0; j < FFN_HIDDEN*H; j++) proj_w[j] += pw[j];
                    for (int j = 0; j < H; j++) proj_b[j] += pb[j];
                    exp_count++;
                }
            }
            if (exp_count > 0) {
                float iv = 1.0f/exp_count;
                for (int j = 0; j < H*FFN_HIDDEN; j++) fc_w[j] *= iv;
                for (int j = 0; j < FFN_HIDDEN; j++) fc_b[j] *= iv;
                for (int j = 0; j < FFN_HIDDEN*H; j++) proj_w[j] *= iv;
                for (int j = 0; j < H; j++) proj_b[j] *= iv;
            }
            
            for (int i = 0; i < seq; i++) {
                // Softmax gate
                float mv = gs[i*N_EXP]; for (int e = 1; e < N_EXP; e++) if (gs[i*N_EXP+e] > mv) mv = gs[i*N_EXP+e];
                float sum = 0; for (int e = 0; e < N_EXP; e++) { gs[i*N_EXP+e] = expf(gs[i*N_EXP+e]-mv); sum += gs[i*N_EXP+e]; }
                for (int e = 0; e < N_EXP; e++) gs[i*N_EXP+e] /= sum;
                
                // Top-3
                int top_i[TOP_K]; float top_v[TOP_K];
                for (int k = 0; k < TOP_K; k++) { top_i[k] = 0; top_v[k] = -1e9f; }
                for (int e = 0; e < N_EXP; e++) {
                    for (int k = 0; k < TOP_K; k++) {
                        if (gs[i*N_EXP+e] > top_v[k]) {
                            for (int k2 = TOP_K-1; k2 > k; k2--) { top_i[k2] = top_i[k2-1]; top_v[k2] = top_v[k2-1]; }
                            top_i[k] = e; top_v[k] = gs[i*N_EXP+e]; break;
                        }
                    }
                }
                
                float moe_out[H] = {0};
                for (int k = 0; k < TOP_K; k++) {
                    float fc[FFN_HIDDEN];
                    for (int j = 0; j < FFN_HIDDEN; j++) {
                        float s = fc_b[j]; for (int d = 0; d < H; d++) s += buf[i*H+d] * fc_w[j*H+d];
                        fc[j] = gelu(s);
                    }
                    float proj[H];
                    for (int j = 0; j < H; j++) {
                        float s = proj_b[j]; for (int d = 0; d < FFN_HIDDEN; d++) s += fc[d] * proj_w[j*FFN_HIDDEN+d];
                        proj[j] = s;
                    }
                    for (int j = 0; j < H; j++) moe_out[j] += top_v[k] * proj[j];
                }
                for (int j = 0; j < H; j++) x[i*H+j] += moe_out[j];
            }
        }
    }
    
    // Final LN + lm_head
    float *ln_f_w = get_w(m, "ln_f.weight");
    float *ln_f_b = get_w(m, "ln_f.bias");
    float *lm_head_w = get_w(m, "lm_head.weight");
    
    for (int i = 0; i < seq; i++) layer_norm(&x[i*H], ln_f_w, ln_f_b, H);
    
    // Sample last token
    float logits[VOCAB];
    for (int v = 0; v < VOCAB; v++) {
        float s = 0;
        for (int j = 0; j < H; j++) s += x[(seq-1)*H+j] * lm_head_w[v*H+j];
        logits[v] = s;
    }
    
    int next = 0; for (int v = 1; v < VOCAB; v++) if (logits[v] > logits[next]) next = v;
    printf("  next token: %d\n", next);
}

int main(int argc, char **argv) {
    printf("Albert-MoE-13 C Inference\n");
    Model m = load_model(
        "/home/bcloud/models/albert-moe/albert_v3.0.best.safetensors",
        "/tmp/albert_idx.bin"
    );
    if (!m.count) return 1;
    
    int tokens[MAX_SEQ] = {1, 100, 101}; // BOS + dummy
    int seq = 3;
    
    clock_t t0 = clock();
    float x[MAX_SEQ * H];
    forward(x, seq, &m, tokens);
    float ms = (float)(clock() - t0) / CLOCKS_PER_SEC * 1000;
    printf("Forward: %.0fms\n", ms);
    
    free(m.raw); free(m.t);
    return 0;
}

File: tools/run_albert.py
#!/usr/bin/env python3
"""Albert-MoE-13 inference — loads safetensors into PyTorch and generates text."""
import torch
import torch.nn as nn
import torch.nn.functional as F
from safetensors import safe_open
import numpy as np
import os, sys, json, time

MODEL_DIR = os.path.expanduser("~/models/albert-moe")
DEVICE = "cpu"

# ── Config ──
with open(os.path.join(MODEL_DIR, "config.json")) as f:
    cfg = json.load(f)
H = cfg["hidden_size"]  # 256
N_LAYERS = cfg["num_hidden_layers"]  # 22 (blocks)
N_HEADS = cfg["num_attention_heads"]  # 4
N_EXP = cfg["num_experts"]  # 12
TOP_K = cfg["num_experts_per_tok"]  # 3
VOCAB = cfg["vocab_size"]  # 32000
HEAD_DIM = H // N_HEADS  # 64

# ── Model ──
class AlbertMoE(nn.Module):
    def __init__(self):
        super().__init__()
        self.embed = nn.Embedding(VOCAB, H)
        self.ln_f = nn.LayerNorm(H)
        self.lm_head = nn.Linear(H, VOCAB, bias=False)
        
        # Attention
        self.q_proj = nn.Linear(H, H, bias=False)
        self.k_proj = nn.Linear(H, H, bias=False)
        self.v_proj = nn.Linear(H, H, bias=False)
        self.o_proj = nn.Linear(H, H, bias=False)
        
        # MoE gate
        self.moe_gate = nn.Linear(H, N_EXP, bias=False)
        
        # Experts: N_EXP × (c_fc + c_proj)
        self.expert_fc = nn.Linear(H, H*4, bias=False)  # c_fc: 256→1024
        self.expert_proj = nn.Linear(H*4, H, bias=False)  # c_proj: 1024→256
        
    def forward(self, x):
        # x: [B, T, H]
        B, T, _ = x.shape
        
        # Self-attention
        q = self.q_proj(x).view(B, T, N_HEADS, HEAD_DIM).transpose(1,2)
        k = self.k_proj(x).view(B, T, N_HEADS, HEAD_DIM).transpose(1,2)
        v = self.v_proj(x).view(B, T, N_HEADS, HEAD_DIM).transpose(1,2)
        attn = F.scaled_dot_product_attention(q, k, v)
        x = x + self.o_proj(attn.transpose(1,2).reshape(B, T, H))
        
        # MoE
        gate_logits = self.moe_gate(x)  # [B, T, 12]
        gate_probs = F.softmax(gate_logits, dim=-1)
        topk_probs, topk_idx = torch.topk(gate_probs, TOP_K, dim=-1)
        
        moe_out = torch.zeros_like(x)
        for b in range(B):
            for t in range(T):
                for k in range(TOP_K):
                    e = topk_idx[b, t, k].item()
                    w = topk_probs[b, t, k]
                    hidden = x[b, t]
                    fc_out = F.gelu(self.expert_fc(hidden))
                    expert_out = self.expert_proj(fc_out)
                    moe_out[b, t] += w * expert_out
        x = x + moe_out
        return x

def load_weights(model, path):
    """Load safetensors into the model."""
    state = {}
    with safe_open(path, framework="pt") as f:
        for key in f.keys():
            state[key] = f.get_tensor(key)
    
    # Map to model parameters
    model.embed.weight.data = state["embed.weight"][:VOCAB]
    model.ln_f.weight.data = state.get("ln_f.weight", torch.ones(H))
    model.ln_f.bias.data = state.get("ln_f.bias", torch.zeros(H))
    model.lm_head.weight.data = state.get("lm_head.weight", state["embed.weight"][:VOCAB])
    
    # Use block 0's weights (single-block model)
    # For the full 33-block model, we'd need the proper architecture
    prefix = "blocks.0.stream_a"
    model.q_proj.weight.data = state[f"{prefix}.attn.q_proj.weight"]
    model.k_proj.weight.data = state[f"{prefix}.attn.k_proj.weight"]
    model.v_proj.weight.data = state[f"{prefix}.attn.v_proj.weight"]
    model.o_proj.weight.data = state[f"{prefix}.attn.o_proj.weight"]
    model.moe_gate.weight.data = state[f"{prefix}.moe.gate.weight"]
    
    # Expert weights from block 3 (first expert block)
    expert_weights = {}
    for e in range(N_EXP):
        prefix_e = f"blocks.3.experts.{e}"
        # Average across all expert blocks
        fc_w = state[f"{prefix_e}.c_fc.weight"]
        fc_b = state[f"{prefix_e}.c_fc.bias"]
        proj_w = state[f"{prefix_e}.c_proj.weight"]
        proj_b = state[f"{prefix_e}.c_proj.bias"]
    
    model.expert_fc.weight.data = state["blocks.3.experts.0.c_fc.weight"]
    model.expert_fc.bias.data = state["blocks.3.experts.0.c_fc.bias"]
    model.expert_proj.weight.data = state["blocks.3.experts.0.c_proj.weight"]
    model.expert_proj.bias.data = state["blocks.3.experts.0.c_proj.bias"]
    
    print(f"Loaded {len(state)} tensors")

# ── Tokenizer ──
class BPETokenizer:
    def __init__(self, vocab_path):
        import json
        with open(vocab_path) as f:
            self.vocab = json.load(f)
        self.id_to_token = {v: k for k, v in self.vocab.items()}
    def encode(self, text):
        words = text.strip().split()
        tokens = []
        for w in words:
            w = w.lower()
            if w in self.vocab:
                tokens.append(self.vocab[w])
            else:
                for ch in w:
                    if ch in self.vocab:
                        tokens.append(self.vocab[ch])
        return tokens
    def decode(self, tokens):
        return "".join(self.id_to_token.get(t, "?") for t in tokens)

# ── Main ──
if __name__ == "__main__":
    print("Loading Albert-MoE-13...")
    model = AlbertMoE()
    load_weights(model, os.path.join(MODEL_DIR, "albert_v3.0.best.safetensors"))
    model.eval()
    
    tok_path = os.path.join(MODEL_DIR, "vocab.json")
    tokenizer = BPETokenizer(tok_path)
    
    prompt = sys.argv[1] if len(sys.argv) > 1 else "Hello"
    tokens = tokenizer.encode(prompt)
    print(f"Prompt: '{prompt}' → {tokens}")
    
    x = torch.tensor([tokens], dtype=torch.long)
    with torch.no_grad():
        t0 = time.time()
        for i in range(8):
            logits = model(x)
            next_token = logits[0, -1].argmax().item()
            x = torch.cat([x, torch.tensor([[next_token]])], dim=1)
            if next_token == 2:  # EOS
                break
        dt = time.time() - t0
    
    output = tokenizer.decode(x[0].tolist())
    print(f"Output: {output}")
    print(f"Time: {dt*1000:.0f}ms for {x.shape[1]} tokens")
